// @thirdweb-dev/wallets/evm/connectors/embedded-wallet/implementations/dist/thirdweb-dev-wallets-evm-connectors-embedded-wallet-implementations.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var embeddedWallet = require('../../../../../dist/embedded-wallet-5fe773c1.cjs.dev.js');
var supportedSmsCountries = require('../../../../../dist/supported-sms-countries-0545c8a6.cjs.dev.js');
require('../../../../../dist/defineProperty-9051a5d3.cjs.dev.js');
require('@paperxyz/sdk-common-utilities');
require('ethers');
require('ethers/lib/utils');
require('@thirdweb-dev/sdk');



exports.AUTH_TOKEN_LOCAL_STORAGE_NAME = embeddedWallet.AUTH_TOKEN_LOCAL_STORAGE_NAME;
exports.AuthProvider = embeddedWallet.AuthProvider;
exports.DEVICE_SHARE_LOCAL_STORAGE_NAME = embeddedWallet.DEVICE_SHARE_LOCAL_STORAGE_NAME;
exports.DEVICE_SHARE_LOCAL_STORAGE_NAME_DEPRECATED = embeddedWallet.DEVICE_SHARE_LOCAL_STORAGE_NAME_DEPRECATED;
exports.EmbeddedWalletSdk = embeddedWallet.EmbeddedWalletSdk;
exports.RecoveryShareManagement = embeddedWallet.RecoveryShareManagement;
exports.UserWalletStatus = embeddedWallet.UserWalletStatus;
exports.WALLET_USER_DETAILS_LOCAL_STORAGE_NAME = embeddedWallet.WALLET_USER_DETAILS_LOCAL_STORAGE_NAME;
exports.WALLET_USER_ID_LOCAL_STORAGE_NAME = embeddedWallet.WALLET_USER_ID_LOCAL_STORAGE_NAME;
exports.supportedSmsCountries = supportedSmsCountries.supportedSmsCountries;
