// thirdweb/dist/esm/extensions/erc7702/__generated__/MinimalAccount/write/createSessionWithSig.js
import { prepareContractCall } from "../../../../../transaction/prepare-contract-call.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
import { once } from "../../../../../utils/promise/once.js";
export const FN_SELECTOR = "0xb5051648";
const FN_INPUTS = [
    {
        components: [
            {
                name: "signer",
                type: "address",
            },
            {
                name: "isWildcard",
                type: "bool",
            },
            {
                name: "expiresAt",
                type: "uint256",
            },
            {
                components: [
                    {
                        name: "target",
                        type: "address",
                    },
                    {
                        name: "selector",
                        type: "bytes4",
                    },
                    {
                        name: "maxValuePerUse",
                        type: "uint256",
                    },
                    {
                        components: [
                            {
                                name: "limitType",
                                type: "uint8",
                            },
                            {
                                name: "limit",
                                type: "uint256",
                            },
                            {
                                name: "period",
                                type: "uint256",
                            },
                        ],
                        name: "valueLimit",
                        type: "tuple",
                    },
                    {
                        components: [
                            {
                                name: "condition",
                                type: "uint8",
                            },
                            {
                                name: "index",
                                type: "uint64",
                            },
                            {
                                name: "refValue",
                                type: "bytes32",
                            },
                            {
                                components: [
                                    {
                                        name: "limitType",
                                        type: "uint8",
                                    },
                                    {
                                        name: "limit",
                                        type: "uint256",
                                    },
                                    {
                                        name: "period",
                                        type: "uint256",
                                    },
                                ],
                                name: "limit",
                                type: "tuple",
                            },
                        ],
                        name: "constraints",
                        type: "tuple[]",
                    },
                ],
                name: "callPolicies",
                type: "tuple[]",
            },
            {
                components: [
                    {
                        name: "target",
                        type: "address",
                    },
                    {
                        name: "maxValuePerUse",
                        type: "uint256",
                    },
                    {
                        components: [
                            {
                                name: "limitType",
                                type: "uint8",
                            },
                            {
                                name: "limit",
                                type: "uint256",
                            },
                            {
                                name: "period",
                                type: "uint256",
                            },
                        ],
                        name: "valueLimit",
                        type: "tuple",
                    },
                ],
                name: "transferPolicies",
                type: "tuple[]",
            },
            {
                name: "uid",
                type: "bytes32",
            },
        ],
        name: "sessionSpec",
        type: "tuple",
    },
    {
        name: "signature",
        type: "bytes",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `createSessionWithSig` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `createSessionWithSig` method is supported.
 * @extension ERC7702
 * @example
 * ```ts
 * import { isCreateSessionWithSigSupported } from "thirdweb/extensions/erc7702";
 *
 * const supported = isCreateSessionWithSigSupported(["0x..."]);
 * ```
 */
export function isCreateSessionWithSigSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "createSessionWithSig" function.
 * @param options - The options for the createSessionWithSig function.
 * @returns The encoded ABI parameters.
 * @extension ERC7702
 * @example
 * ```ts
 * import { encodeCreateSessionWithSigParams } from "thirdweb/extensions/erc7702";
 * const result = encodeCreateSessionWithSigParams({
 *  sessionSpec: ...,
 *  signature: ...,
 * });
 * ```
 */
export function encodeCreateSessionWithSigParams(options) {
    return encodeAbiParameters(FN_INPUTS, [
        options.sessionSpec,
        options.signature,
    ]);
}
/**
 * Encodes the "createSessionWithSig" function into a Hex string with its parameters.
 * @param options - The options for the createSessionWithSig function.
 * @returns The encoded hexadecimal string.
 * @extension ERC7702
 * @example
 * ```ts
 * import { encodeCreateSessionWithSig } from "thirdweb/extensions/erc7702";
 * const result = encodeCreateSessionWithSig({
 *  sessionSpec: ...,
 *  signature: ...,
 * });
 * ```
 */
export function encodeCreateSessionWithSig(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeCreateSessionWithSigParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "createSessionWithSig" function on the contract.
 * @param options - The options for the "createSessionWithSig" function.
 * @returns A prepared transaction object.
 * @extension ERC7702
 * @example
 * ```ts
 * import { sendTransaction } from "thirdweb";
 * import { createSessionWithSig } from "thirdweb/extensions/erc7702";
 *
 * const transaction = createSessionWithSig({
 *  contract,
 *  sessionSpec: ...,
 *  signature: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * await sendTransaction({ transaction, account });
 * ```
 */
export function createSessionWithSig(options) {
    const asyncOptions = once(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return prepareContractCall({
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        authorizationList: async () => (await asyncOptions()).overrides?.authorizationList,
        contract: options.contract,
        erc20Value: async () => (await asyncOptions()).overrides?.erc20Value,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.sessionSpec, resolvedOptions.signature];
        },
        value: async () => (await asyncOptions()).overrides?.value,
    });
}
