// thirdweb/dist/esm/react/core/hooks/pay/useBuyWithFiatQuotesForProviders.js
import { useQueries } from "@tanstack/react-query";
import { prepare as prepareOnramp } from "../../../../bridge/Onramp.js";
import { getToken } from "../../../../pay/convert/get-token.js";
import { toUnits } from "../../../../utils/units.js";
/**
 * @internal
 * Hook to get prepared onramp quotes from Coinbase, Stripe, and Transak providers.
 */
export function useBuyWithFiatQuotesForProviders(params, queryOptions) {
    const providers = ["coinbase", "stripe", "transak"];
    const queries = useQueries({
        queries: providers.map((provider) => ({
            ...queryOptions,
            enabled: !!params,
            queryFn: async () => {
                if (!params) {
                    throw new Error("No params provided");
                }
                const token = await getToken(params.client, params.tokenAddress, params.chainId);
                const amountWei = toUnits(params.amount, token.decimals);
                return prepareOnramp({
                    amount: amountWei,
                    chainId: params.chainId,
                    client: params.client,
                    currency: params.currency || "USD",
                    onramp: provider,
                    receiver: params.receiver,
                    tokenAddress: params.tokenAddress,
                });
            },
            queryKey: ["onramp-prepare", provider, params],
            retry: false,
        })),
    });
    return queries;
}
