// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/swap/StepConnector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepConnectorArrow = StepConnectorArrow;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_icons_1 = require("@radix-ui/react-icons");
const CustomThemeProvider_js_1 = require("../../../../../../core/design-system/CustomThemeProvider.js");
const basic_js_1 = require("../../../../components/basic.js");
function StepConnectorArrow() {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { center: "both", flex: "row", style: {
            marginBottom: "-10px",
            marginTop: "-10px",
            position: "relative",
            width: "100%",
            zIndex: 1000,
        }, children: (0, jsx_runtime_1.jsx)(basic_js_1.Container, { center: "both", color: "primaryText", flex: "row", style: {
                backgroundColor: theme.colors.modalBg,
                border: `1px solid ${theme.colors.borderColor}`,
                borderRadius: "100%",
                height: "30px",
                width: "30px",
            }, children: (0, jsx_runtime_1.jsx)(react_icons_1.ChevronDownIcon, { height: 16, width: 16 }) }) }));
}
