// thirdweb/dist/cjs/stories/Bridge/FundWallet.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomButtonLabelLight = exports.CustomButtonLabel = exports.LargeAmountLight = exports.LargeAmount = exports.USDCTokenLight = exports.USDCToken = exports.WithInitialAmountLight = exports.WithInitialAmount = exports.Dark = exports.Light = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const FundWallet_js_1 = require("../../react/web/ui/Bridge/FundWallet.js");
const utils_js_1 = require("../utils.js");
const fixtures_js_1 = require("./fixtures.js");
// Wrapper component to provide theme context
const FundWalletWithTheme = (props) => {
    const { theme, ...componentProps } = props;
    return ((0, jsx_runtime_1.jsx)(utils_js_1.ModalThemeWrapper, { theme: theme, children: (0, jsx_runtime_1.jsx)(FundWallet_js_1.FundWallet, { ...componentProps }) }));
};
const meta = {
    args: {
        client: utils_js_1.storyClient,
        onContinue: (amount, token, receiverAddress) => {
            alert(`Continue with ${amount} ${token.symbol} to ${receiverAddress}`);
        },
        receiverAddress: fixtures_js_1.RECEIVER_ADDRESSES.primary,
        theme: "dark",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.ethDefault,
    },
    argTypes: {
        onContinue: { action: "continue clicked" },
        receiverAddress: {
            description: "Optional receiver address (defaults to connected wallet)",
        },
        theme: {
            control: "select",
            description: "Theme for the component",
            options: ["light", "dark"],
        },
        uiOptions: {
            description: "UI configuration for fund wallet mode",
        },
    },
    component: FundWalletWithTheme,
    parameters: {
        docs: {
            description: {
                component: "FundWallet component allows users to specify the amount they want to add to their wallet. This is the first screen in the fund_wallet flow before method selection.\n\n" +
                    "## Features\n" +
                    "- **Token Selection**: Choose from different tokens (ETH, USDC, UNI)\n" +
                    "- **Amount Input**: Enter custom amount or use quick options\n" +
                    "- **Receiver Address**: Optional receiver address (defaults to connected wallet)\n" +
                    "- **Quick Options**: Preset amounts for faster selection\n" +
                    "- **Theme Support**: Works with both light and dark themes\n\n" +
                    "This component now accepts uiOptions directly to configure the destination token, initial amount, and quick options.",
            },
        },
        layout: "centered",
    },
    tags: ["autodocs"],
    title: "Bridge/FundWallet",
};
exports.default = meta;
exports.Light = {
    args: {
        receiverAddress: undefined,
        theme: "light",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.ethDefault,
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "Default fund wallet interface in light theme with ETH token.",
            },
        },
    },
};
exports.Dark = {
    args: {
        receiverAddress: undefined,
        theme: "dark",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.ethDefault,
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Default fund wallet interface in dark theme with ETH token.",
            },
        },
    },
};
exports.WithInitialAmount = {
    args: {
        receiverAddress: fixtures_js_1.RECEIVER_ADDRESSES.secondary,
        theme: "dark",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.ethWithAmount,
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Fund wallet with pre-filled amount and specified receiver address.",
            },
        },
    },
};
exports.WithInitialAmountLight = {
    args: {
        receiverAddress: fixtures_js_1.RECEIVER_ADDRESSES.secondary,
        theme: "light",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.ethWithAmount,
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "Light theme version with pre-filled amount and receiver address.",
            },
        },
    },
};
exports.USDCToken = {
    args: {
        theme: "dark",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.usdcDefault,
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Fund wallet configured for USDC token with initial amount.",
            },
        },
    },
};
exports.USDCTokenLight = {
    args: {
        theme: "light",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.usdcDefault,
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "Light theme version for USDC token funding.",
            },
        },
    },
};
exports.LargeAmount = {
    args: {
        theme: "dark",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.uniLarge,
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Fund wallet with UNI token and large pre-filled amount to test formatting.",
            },
        },
    },
};
exports.LargeAmountLight = {
    args: {
        theme: "light",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.uniLarge,
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "Light theme version with UNI token and large amount.",
            },
        },
    },
};
exports.CustomButtonLabel = {
    args: {
        theme: "dark",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.customButton,
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Example showcasing custom button label functionality. The button shows 'Add Funds Now' instead of the default 'Buy [amount] [symbol]' text.",
            },
        },
    },
};
exports.CustomButtonLabelLight = {
    args: {
        theme: "light",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.customButton,
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "Light theme version with custom button label 'Add Funds Now'.",
            },
        },
    },
};
