// thirdweb/dist/cjs/stories/Bridge/StepRunner.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dark = exports.Light = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const StepRunner_js_1 = require("../../react/web/ui/Bridge/StepRunner.js");
const utils_js_1 = require("../utils.js");
const fixtures_js_1 = require("./fixtures.js");
// Mock window adapter
const mockWindowAdapter = {
    open: async (_url) => { },
};
// Wrapper component to provide theme context
const StepRunnerWithTheme = (props) => {
    const { theme, ...componentProps } = props;
    return ((0, jsx_runtime_1.jsx)(utils_js_1.ModalThemeWrapper, { theme: theme, children: (0, jsx_runtime_1.jsx)(StepRunner_js_1.StepRunner, { ...componentProps }) }));
};
const meta = {
    args: {
        client: utils_js_1.storyClient,
        onCancel: () => { },
        onComplete: (_completedStatuses) => { },
        onError: (error) => console.error("Error:", error),
        theme: "dark",
        wallet: fixtures_js_1.STORY_MOCK_WALLET,
        windowAdapter: mockWindowAdapter,
    },
    argTypes: {
        onCancel: { action: "execution cancelled" },
        onComplete: { action: "execution completed" },
        onError: { action: "error occurred" },
        theme: {
            control: "select",
            description: "Theme for the component",
            options: ["light", "dark"],
        },
    },
    component: StepRunnerWithTheme,
    parameters: {
        docs: {
            description: {
                component: "**StepRunner** executes prepared route steps sequentially, showing real-time progress and transaction status.\n\n" +
                    "## Features\n" +
                    "- **Real Execution**: Uses useStepExecutor hook for actual transaction processing\n" +
                    "- **Progress Tracking**: Visual progress bar and step-by-step status updates\n" +
                    "- **Error Handling**: Retry functionality for failed transactions\n" +
                    "- **Transaction Batching**: Optimizes multiple transactions when possible\n" +
                    "- **Onramp Support**: Handles fiat-to-crypto onramp flows\n\n" +
                    "## Props\n" +
                    "- `steps`: Array of RouteStep objects from Bridge.prepare()\n" +
                    "- `wallet`: Connected wallet for transaction signing\n" +
                    "- `client`: ThirdwebClient instance\n" +
                    "- `windowAdapter`: Platform-specific window/URL handler\n" +
                    "- `onramp`: Optional onramp configuration\n\n" +
                    "## Integration\n" +
                    "This component is typically used within the BridgeOrchestrator after route preparation.",
            },
        },
        layout: "centered",
    },
    tags: ["autodocs"],
    title: "Bridge/StepRunner",
};
exports.default = meta;
exports.Light = {
    args: {
        request: fixtures_js_1.simpleBuyRequest,
        theme: "light",
    },
    parameters: {
        backgrounds: { default: "light" },
    },
};
exports.Dark = {
    args: {
        request: fixtures_js_1.simpleBuyRequest,
        theme: "dark",
    },
    parameters: {
        backgrounds: { default: "dark" },
    },
};
