// thirdweb/dist/cjs/stories/Bridge/SuccessScreen.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentId = exports.TransactionPayment = exports.ComplexPaymentLight = exports.ComplexPayment = exports.OnrampPaymentLight = exports.OnrampPayment = exports.DefaultLight = exports.Default = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const viem_1 = require("viem");
const WindowAdapter_js_1 = require("../../react/web/adapters/WindowAdapter.js");
const SuccessScreen_js_1 = require("../../react/web/ui/Bridge/payment-success/SuccessScreen.js");
const utils_js_1 = require("../utils.js");
const fixtures_js_1 = require("./fixtures.js");
const mockBuyCompletedStatuses = JSON.parse((0, viem_1.stringify)([
    {
        destinationAmount: 100000000n,
        destinationChainId: 1,
        destinationToken: {
            address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
            chainId: 1,
            decimals: 6,
            name: "USD Coin",
            priceUsd: 1,
            symbol: "USDC",
        },
        destinationTokenAddress: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
        originAmount: 1000000000000000000n,
        originChainId: 1,
        originToken: {
            address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
            chainId: 1,
            decimals: 18,
            name: "Ethereum",
            priceUsd: 2500,
            symbol: "ETH",
        },
        originTokenAddress: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
        paymentId: "payment-12345",
        receiver: "0xa3841994009B4fEabb01ebcC62062F9E56F701CD",
        sender: "0xa3841994009B4fEabb01ebcC62062F9E56F701CD",
        status: "COMPLETED",
        transactions: [
            {
                chainId: 1,
                transactionHash: "0x1234567890abcdef1234567890abcdef12345678901234567890abcdef123456",
            },
        ],
        type: "buy",
    },
]));
const mockOnrampCompletedStatuses = JSON.parse((0, viem_1.stringify)([
    {
        purchaseData: {
            orderId: "stripe-order-abc123",
        },
        status: "COMPLETED",
        transactions: [
            {
                chainId: 137,
                transactionHash: "0xabcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890",
            },
        ],
        type: "onramp",
    },
]));
// Wrapper component to provide theme context
const SuccessScreenWithTheme = (props) => {
    const { theme, ...componentProps } = props;
    return ((0, jsx_runtime_1.jsx)(utils_js_1.ModalThemeWrapper, { theme: theme, children: (0, jsx_runtime_1.jsx)(SuccessScreen_js_1.SuccessScreen, { ...componentProps }) }));
};
const meta = {
    args: {
        completedStatuses: mockBuyCompletedStatuses,
        onDone: () => { },
        preparedQuote: fixtures_js_1.simpleBuyQuote,
        theme: "dark",
        uiOptions: fixtures_js_1.FUND_WALLET_UI_OPTIONS.ethDefault,
        windowAdapter: WindowAdapter_js_1.webWindowAdapter,
    },
    argTypes: {
        onDone: { action: "success screen closed" },
        theme: {
            control: "select",
            description: "Theme for the component",
            options: ["light", "dark"],
        },
    },
    component: SuccessScreenWithTheme,
    parameters: {
        docs: {
            description: {
                component: "Success screen that displays completion confirmation with transaction summary, payment details, and action buttons for next steps. Includes animated success icon and detailed transaction view.",
            },
        },
        layout: "centered",
    },
    tags: ["autodocs"],
    title: "Bridge/SuccessScreen",
};
exports.default = meta;
exports.Default = {
    args: {
        client: utils_js_1.storyClient,
        theme: "dark",
    },
    parameters: {
        backgrounds: { default: "dark" },
    },
};
exports.DefaultLight = {
    args: {
        client: utils_js_1.storyClient,
        theme: "light",
    },
    parameters: {
        backgrounds: { default: "light" },
    },
};
exports.OnrampPayment = {
    args: {
        client: utils_js_1.storyClient,
        completedStatuses: mockOnrampCompletedStatuses,
        preparedQuote: fixtures_js_1.simpleOnrampQuote,
        theme: "dark",
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Success screen for onramp payments showing payment ID that can be copied to clipboard.",
            },
        },
    },
};
exports.OnrampPaymentLight = {
    args: {
        client: utils_js_1.storyClient,
        completedStatuses: mockOnrampCompletedStatuses,
        preparedQuote: fixtures_js_1.simpleOnrampQuote,
        theme: "light",
    },
    parameters: {
        backgrounds: { default: "light" },
    },
};
exports.ComplexPayment = {
    args: {
        client: utils_js_1.storyClient,
        completedStatuses: [
            ...mockOnrampCompletedStatuses,
            ...mockBuyCompletedStatuses,
        ],
        preparedQuote: fixtures_js_1.simpleOnrampQuote,
        theme: "dark",
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Success screen for onramp payments showing payment ID that can be copied to clipboard.",
            },
        },
    },
};
exports.ComplexPaymentLight = {
    args: {
        client: utils_js_1.storyClient,
        completedStatuses: [
            ...mockOnrampCompletedStatuses,
            ...mockBuyCompletedStatuses,
        ],
        preparedQuote: fixtures_js_1.simpleOnrampQuote,
        theme: "light",
    },
    parameters: {
        backgrounds: { default: "light" },
    },
};
exports.TransactionPayment = {
    args: {
        client: utils_js_1.storyClient,
        completedStatuses: mockBuyCompletedStatuses,
        preparedQuote: fixtures_js_1.simpleBuyQuote,
        theme: "light",
        uiOptions: fixtures_js_1.TRANSACTION_UI_OPTIONS.contractInteraction,
    },
    parameters: {
        backgrounds: { default: "light" },
    },
};
exports.PaymentId = {
    args: {
        client: utils_js_1.storyClient,
        completedStatuses: mockBuyCompletedStatuses,
        hasPaymentId: true,
        preparedQuote: fixtures_js_1.simpleBuyQuote,
        theme: "light",
        uiOptions: fixtures_js_1.TRANSACTION_UI_OPTIONS.contractInteraction,
    },
    parameters: {
        backgrounds: { default: "light" },
    },
};
