// thirdweb/dist/cjs/stories/Bridge/UnsupportedTokenScreen.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestnetNotSupportedLight = exports.TestnetNotSupported = exports.TokenNotSupportedLight = exports.TokenNotSupported = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const utils_js_1 = require("../../chains/utils.js");
const client_js_1 = require("../../client/client.js");
const UnsupportedTokenScreen_js_1 = require("../../react/web/ui/Bridge/UnsupportedTokenScreen.js");
const utils_js_2 = require("../utils.js");
const TEST_CLIENT = (0, client_js_1.createThirdwebClient)({ clientId: "test" });
// Wrapper component to provide theme context
const UnsupportedTokenScreenWithTheme = (props) => {
    const { theme, ...componentProps } = props;
    return ((0, jsx_runtime_1.jsx)(utils_js_2.ModalThemeWrapper, { theme: theme, children: (0, jsx_runtime_1.jsx)(UnsupportedTokenScreen_js_1.UnsupportedTokenScreen, { ...componentProps }) }));
};
const meta = {
    args: {
        chain: (0, utils_js_1.defineChain)(1), // Ethereum mainnet
        theme: "dark",
    },
    argTypes: {
        theme: {
            control: "select",
            description: "Theme for the component",
            options: ["light", "dark"],
        },
    },
    component: UnsupportedTokenScreenWithTheme,
    parameters: {
        docs: {
            description: {
                component: "Screen displayed when a token is being indexed or when using an unsupported testnet. Shows loading state for indexing tokens or error state for testnets.",
            },
        },
        layout: "centered",
    },
    tags: ["autodocs"],
    title: "Bridge/UnsupportedTokenScreen",
};
exports.default = meta;
exports.TokenNotSupported = {
    args: {
        chain: (0, utils_js_1.defineChain)(1),
        client: TEST_CLIENT,
        theme: "dark", // Ethereum mainnet - will show indexing spinner
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Shows the loading state when a token is being indexed by the Universal Bridge on a mainnet chain.",
            },
        },
    },
};
exports.TokenNotSupportedLight = {
    args: {
        chain: (0, utils_js_1.defineChain)(1),
        client: TEST_CLIENT,
        theme: "light", // Ethereum mainnet - will show indexing spinner
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "Shows the loading state when a token is being indexed by the Universal Bridge on a mainnet chain (light theme).",
            },
        },
    },
};
exports.TestnetNotSupported = {
    args: {
        chain: (0, utils_js_1.defineChain)(11155111),
        client: TEST_CLIENT,
        theme: "dark", // Sepolia testnet - will show error state
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Shows the error state when trying to use the Universal Bridge on a testnet chain (Sepolia in this example).",
            },
        },
    },
};
exports.TestnetNotSupportedLight = {
    args: {
        chain: (0, utils_js_1.defineChain)(11155111),
        client: TEST_CLIENT,
        theme: "light", // Sepolia testnet - will show error state
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "Shows the error state when trying to use the Universal Bridge on a testnet chain (Sepolia in this example, light theme).",
            },
        },
    },
};
