// @thirdweb-dev/react-core/evm/dist/thirdweb-dev-react-core-evm.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var useBuyWithCryptoStatus = require('../../dist/useBuyWithCryptoStatus-768f0ea5.cjs.prod.js');
require('mime/lite.js');
require('react');
require('@tanstack/react-query');
require('tiny-invariant');
require('react/jsx-runtime');
require('@thirdweb-dev/chains');
require('@thirdweb-dev/sdk/evm');
require('@thirdweb-dev/sdk');
require('@thirdweb-dev/wallets');
require('ethers');
require('@thirdweb-dev/auth');
require('@thirdweb-dev/sdk/evm/functions');



exports.ThirdwebAuthProvider = useBuyWithCryptoStatus.ThirdwebAuthProvider;
exports.ThirdwebConfigProvider = useBuyWithCryptoStatus.ThirdwebConfigProvider;
exports.ThirdwebConnectedWalletProvider = useBuyWithCryptoStatus.ThirdwebConnectedWalletProvider;
exports.ThirdwebProviderCore = useBuyWithCryptoStatus.ThirdwebProviderCore;
exports.ThirdwebSDKProvider = useBuyWithCryptoStatus.ThirdwebSDKProvider;
exports.ThirdwebThemeContext = useBuyWithCryptoStatus.ThirdwebThemeContext;
exports.__DEV__ = useBuyWithCryptoStatus.__DEV__;
exports.compilerMetadata = useBuyWithCryptoStatus.compilerMetadata;
exports.contractType = useBuyWithCryptoStatus.contractType;
exports.getErc1155 = useBuyWithCryptoStatus.getErc1155;
exports.getErc20 = useBuyWithCryptoStatus.getErc20;
exports.getErc721 = useBuyWithCryptoStatus.getErc721;
exports.getErcs = useBuyWithCryptoStatus.getErcs;
exports.invalidateContractAndBalances = useBuyWithCryptoStatus.invalidateContractAndBalances;
exports.isEnsName = useBuyWithCryptoStatus.isEnsName;
exports.isPossibleEVMAddress = useBuyWithCryptoStatus.isPossibleEVMAddress;
exports.resolveMimeType = useBuyWithCryptoStatus.resolveMimeType;
exports.shortenAddress = useBuyWithCryptoStatus.shortenAddress;
exports.shortenIfAddress = useBuyWithCryptoStatus.shortenIfAddress;
exports.shortenString = useBuyWithCryptoStatus.shortenString;
exports.shouldNeverPersistQuery = useBuyWithCryptoStatus.shouldNeverPersistQuery;
exports.useAcceptDirectListingOffer = useBuyWithCryptoStatus.useAcceptDirectListingOffer;
exports.useAccountAdmins = useBuyWithCryptoStatus.useAccountAdmins;
exports.useAccountAdminsAndSigners = useBuyWithCryptoStatus.useAccountAdminsAndSigners;
exports.useAccountSigners = useBuyWithCryptoStatus.useAccountSigners;
exports.useAccounts = useBuyWithCryptoStatus.useAccounts;
exports.useAccountsForAddress = useBuyWithCryptoStatus.useAccountsForAddress;
exports.useActiveChain = useBuyWithCryptoStatus.useActiveChain;
exports.useActiveClaimCondition = useBuyWithCryptoStatus.useActiveClaimCondition;
exports.useActiveClaimConditionForWallet = useBuyWithCryptoStatus.useActiveClaimConditionForWallet;
exports.useActiveListings = useBuyWithCryptoStatus.useActiveListings;
exports.useAddAdmin = useBuyWithCryptoStatus.useAddAdmin;
exports.useAddress = useBuyWithCryptoStatus.useAddress;
exports.useAirdropNFT = useBuyWithCryptoStatus.useAirdropNFT;
exports.useAllRoleMembers = useBuyWithCryptoStatus.useAllRoleMembers;
exports.useAppURI = useBuyWithCryptoStatus.useAppURI;
exports.useAuctionWinner = useBuyWithCryptoStatus.useAuctionWinner;
exports.useAuth = useBuyWithCryptoStatus.useAuth;
exports.useBalance = useBuyWithCryptoStatus.useBalance;
exports.useBalanceForAddress = useBuyWithCryptoStatus.useBalanceForAddress;
exports.useBatchesToReveal = useBuyWithCryptoStatus.useBatchesToReveal;
exports.useBidBuffer = useBuyWithCryptoStatus.useBidBuffer;
exports.useBurnNFT = useBuyWithCryptoStatus.useBurnNFT;
exports.useBurnToken = useBuyWithCryptoStatus.useBurnToken;
exports.useBuyDirectListing = useBuyWithCryptoStatus.useBuyDirectListing;
exports.useBuyNow = useBuyWithCryptoStatus.useBuyNow;
exports.useBuyWithCryptoHistory = useBuyWithCryptoStatus.useBuyWithCryptoHistory;
exports.useBuyWithCryptoQuote = useBuyWithCryptoStatus.useBuyWithCryptoQuote;
exports.useBuyWithCryptoStatus = useBuyWithCryptoStatus.useBuyWithCryptoStatus;
exports.useCancelDirectListing = useBuyWithCryptoStatus.useCancelDirectListing;
exports.useCancelEnglishAuction = useBuyWithCryptoStatus.useCancelEnglishAuction;
exports.useCancelListing = useBuyWithCryptoStatus.useCancelListing;
exports.useChain = useBuyWithCryptoStatus.useChain;
exports.useChainId = useBuyWithCryptoStatus.useChainId;
exports.useClaimConditions = useBuyWithCryptoStatus.useClaimConditions;
exports.useClaimIneligibilityReasons = useBuyWithCryptoStatus.useClaimIneligibilityReasons;
exports.useClaimNFT = useBuyWithCryptoStatus.useClaimNFT;
exports.useClaimToken = useBuyWithCryptoStatus.useClaimToken;
exports.useClaimedNFTSupply = useBuyWithCryptoStatus.useClaimedNFTSupply;
exports.useClaimedNFTs = useBuyWithCryptoStatus.useClaimedNFTs;
exports.useClaimerProofs = useBuyWithCryptoStatus.useClaimerProofs;
exports.useCompilerMetadata = useBuyWithCryptoStatus.useCompilerMetadata;
exports.useConnect = useBuyWithCryptoStatus.useConnect;
exports.useConnectedWallet = useBuyWithCryptoStatus.useConnectedWallet;
exports.useConnectionStatus = useBuyWithCryptoStatus.useConnectionStatus;
exports.useContract = useBuyWithCryptoStatus.useContract;
exports.useContractEvents = useBuyWithCryptoStatus.useContractEvents;
exports.useContractMetadata = useBuyWithCryptoStatus.useContractMetadata;
exports.useContractMetadataUpdate = useBuyWithCryptoStatus.useContractMetadataUpdate;
exports.useContractRead = useBuyWithCryptoStatus.useContractRead;
exports.useContractType = useBuyWithCryptoStatus.useContractType;
exports.useContractWrite = useBuyWithCryptoStatus.useContractWrite;
exports.useCreateAccount = useBuyWithCryptoStatus.useCreateAccount;
exports.useCreateAuctionListing = useBuyWithCryptoStatus.useCreateAuctionListing;
exports.useCreateDirectListing = useBuyWithCryptoStatus.useCreateDirectListing;
exports.useCreateDirectListingBatch = useBuyWithCryptoStatus.useCreateDirectListingBatch;
exports.useCreateSessionKey = useBuyWithCryptoStatus.useCreateSessionKey;
exports.useCreateWalletInstance = useBuyWithCryptoStatus.useCreateWalletInstance;
exports.useDelayedRevealLazyMint = useBuyWithCryptoStatus.useDelayedRevealLazyMint;
exports.useDirectListing = useBuyWithCryptoStatus.useDirectListing;
exports.useDirectListings = useBuyWithCryptoStatus.useDirectListings;
exports.useDirectListingsCount = useBuyWithCryptoStatus.useDirectListingsCount;
exports.useDisconnect = useBuyWithCryptoStatus.useDisconnect;
exports.useENS = useBuyWithCryptoStatus.useENS;
exports.useEdition = useBuyWithCryptoStatus.useEdition;
exports.useEditionDrop = useBuyWithCryptoStatus.useEditionDrop;
exports.useEnglishAuction = useBuyWithCryptoStatus.useEnglishAuction;
exports.useEnglishAuctionWinningBid = useBuyWithCryptoStatus.useEnglishAuctionWinningBid;
exports.useEnglishAuctions = useBuyWithCryptoStatus.useEnglishAuctions;
exports.useEnglishAuctionsCount = useBuyWithCryptoStatus.useEnglishAuctionsCount;
exports.useExecuteAuctionSale = useBuyWithCryptoStatus.useExecuteAuctionSale;
exports.useGrantRole = useBuyWithCryptoStatus.useGrantRole;
exports.useIsAccountDeployed = useBuyWithCryptoStatus.useIsAccountDeployed;
exports.useIsAddressRole = useBuyWithCryptoStatus.useIsAddressRole;
exports.useLazyMint = useBuyWithCryptoStatus.useLazyMint;
exports.useListing = useBuyWithCryptoStatus.useListing;
exports.useListings = useBuyWithCryptoStatus.useListings;
exports.useListingsCount = useBuyWithCryptoStatus.useListingsCount;
exports.useLogin = useBuyWithCryptoStatus.useLogin;
exports.useLogout = useBuyWithCryptoStatus.useLogout;
exports.useMakeBid = useBuyWithCryptoStatus.useMakeBid;
exports.useMakeOffer = useBuyWithCryptoStatus.useMakeOffer;
exports.useMarketplace = useBuyWithCryptoStatus.useMarketplace;
exports.useMetadata = useBuyWithCryptoStatus.useMetadata;
exports.useMinimumNextBid = useBuyWithCryptoStatus.useMinimumNextBid;
exports.useMintNFT = useBuyWithCryptoStatus.useMintNFT;
exports.useMintNFTSupply = useBuyWithCryptoStatus.useMintNFTSupply;
exports.useMintToken = useBuyWithCryptoStatus.useMintToken;
exports.useMultiwrap = useBuyWithCryptoStatus.useMultiwrap;
exports.useNFT = useBuyWithCryptoStatus.useNFT;
exports.useNFTBalance = useBuyWithCryptoStatus.useNFTBalance;
exports.useNFTCollection = useBuyWithCryptoStatus.useNFTCollection;
exports.useNFTDrop = useBuyWithCryptoStatus.useNFTDrop;
exports.useNFTs = useBuyWithCryptoStatus.useNFTs;
exports.useNetwork = useBuyWithCryptoStatus.useNetwork;
exports.useNetworkMismatch = useBuyWithCryptoStatus.useNetworkMismatch;
exports.useOffers = useBuyWithCryptoStatus.useOffers;
exports.useOwnedNFTs = useBuyWithCryptoStatus.useOwnedNFTs;
exports.usePack = useBuyWithCryptoStatus.usePack;
exports.usePersonalWalletAddress = useBuyWithCryptoStatus.usePersonalWalletAddress;
exports.usePlatformFees = useBuyWithCryptoStatus.usePlatformFees;
exports.usePrimarySaleRecipient = useBuyWithCryptoStatus.usePrimarySaleRecipient;
exports.useReadonlySDK = useBuyWithCryptoStatus.useReadonlySDK;
exports.useRemoveAdmin = useBuyWithCryptoStatus.useRemoveAdmin;
exports.useResetClaimConditions = useBuyWithCryptoStatus.useResetClaimConditions;
exports.useRevealLazyMint = useBuyWithCryptoStatus.useRevealLazyMint;
exports.useRevokeRole = useBuyWithCryptoStatus.useRevokeRole;
exports.useRevokeSessionKey = useBuyWithCryptoStatus.useRevokeSessionKey;
exports.useRoleMembers = useBuyWithCryptoStatus.useRoleMembers;
exports.useRoyaltySettings = useBuyWithCryptoStatus.useRoyaltySettings;
exports.useSDK = useBuyWithCryptoStatus.useSDK;
exports.useSDKChainId = useBuyWithCryptoStatus.useSDKChainId;
exports.useSetAllRoleMembers = useBuyWithCryptoStatus.useSetAllRoleMembers;
exports.useSetAppURI = useBuyWithCryptoStatus.useSetAppURI;
exports.useSetClaimConditions = useBuyWithCryptoStatus.useSetClaimConditions;
exports.useSetConnectedWallet = useBuyWithCryptoStatus.useSetConnectedWallet;
exports.useSetConnectionStatus = useBuyWithCryptoStatus.useSetConnectionStatus;
exports.useSetSharedMetadata = useBuyWithCryptoStatus.useSetSharedMetadata;
exports.useSharedMetadata = useBuyWithCryptoStatus.useSharedMetadata;
exports.useSignatureDrop = useBuyWithCryptoStatus.useSignatureDrop;
exports.useSigner = useBuyWithCryptoStatus.useSigner;
exports.useSplit = useBuyWithCryptoStatus.useSplit;
exports.useStorage = useBuyWithCryptoStatus.useStorage;
exports.useStorageUpload = useBuyWithCryptoStatus.useStorageUpload;
exports.useSupportedChains = useBuyWithCryptoStatus.useSupportedChains;
exports.useSupportedWallet = useBuyWithCryptoStatus.useSupportedWallet;
exports.useSwitchAccount = useBuyWithCryptoStatus.useSwitchAccount;
exports.useSwitchChain = useBuyWithCryptoStatus.useSwitchChain;
exports.useThirdwebAuthContext = useBuyWithCryptoStatus.useThirdwebAuthContext;
exports.useThirdwebConfigContext = useBuyWithCryptoStatus.useThirdwebConfigContext;
exports.useThirdwebConnectedWalletContext = useBuyWithCryptoStatus.useThirdwebConnectedWalletContext;
exports.useToken = useBuyWithCryptoStatus.useToken;
exports.useTokenBalance = useBuyWithCryptoStatus.useTokenBalance;
exports.useTokenDecimals = useBuyWithCryptoStatus.useTokenDecimals;
exports.useTokenDrop = useBuyWithCryptoStatus.useTokenDrop;
exports.useTokenSupply = useBuyWithCryptoStatus.useTokenSupply;
exports.useTotalCirculatingSupply = useBuyWithCryptoStatus.useTotalCirculatingSupply;
exports.useTotalCount = useBuyWithCryptoStatus.useTotalCount;
exports.useTransferBatchToken = useBuyWithCryptoStatus.useTransferBatchToken;
exports.useTransferNFT = useBuyWithCryptoStatus.useTransferNFT;
exports.useTransferNativeToken = useBuyWithCryptoStatus.useTransferNativeToken;
exports.useTransferToken = useBuyWithCryptoStatus.useTransferToken;
exports.useUnclaimedNFTSupply = useBuyWithCryptoStatus.useUnclaimedNFTSupply;
exports.useUnclaimedNFTs = useBuyWithCryptoStatus.useUnclaimedNFTs;
exports.useUpdateMetadata = useBuyWithCryptoStatus.useUpdateMetadata;
exports.useUpdateNFTMetadata = useBuyWithCryptoStatus.useUpdateNFTMetadata;
exports.useUpdatePlatformFees = useBuyWithCryptoStatus.useUpdatePlatformFees;
exports.useUpdatePrimarySaleRecipient = useBuyWithCryptoStatus.useUpdatePrimarySaleRecipient;
exports.useUpdateRoyaltySettings = useBuyWithCryptoStatus.useUpdateRoyaltySettings;
exports.useUser = useBuyWithCryptoStatus.useUser;
exports.useValidDirectListings = useBuyWithCryptoStatus.useValidDirectListings;
exports.useValidEnglishAuctions = useBuyWithCryptoStatus.useValidEnglishAuctions;
exports.useVote = useBuyWithCryptoStatus.useVote;
exports.useWallet = useBuyWithCryptoStatus.useWallet;
exports.useWalletConfig = useBuyWithCryptoStatus.useWalletConfig;
exports.useWalletConnectHandler = useBuyWithCryptoStatus.useWalletConnectHandler;
exports.useWalletContext = useBuyWithCryptoStatus.useWalletContext;
exports.useWallets = useBuyWithCryptoStatus.useWallets;
exports.useWatchTransactions = useBuyWithCryptoStatus.useWatchTransactions;
exports.useWinningBid = useBuyWithCryptoStatus.useWinningBid;
