// thirdweb/dist/esm/exports/extensions/permissions.js
//--------------------------------------------------
// Permissions
// --------------------------------------------------------
export { roleAdminChangedEvent, } from "../../extensions/permissions/__generated__/IPermissions/events/RoleAdminChanged.js";
// EVENTS
export { roleGrantedEvent, } from "../../extensions/permissions/__generated__/IPermissions/events/RoleGranted.js";
export { roleRevokedEvent, } from "../../extensions/permissions/__generated__/IPermissions/events/RoleRevoked.js";
export { getRoleAdmin, isGetRoleAdminSupported, } from "../../extensions/permissions/read/getRoleAdmin.js";
// READ
export { hasRole, isHasRoleSupported, } from "../../extensions/permissions/read/hasRole.js";
// WRITE
export { grantRole, isGrantRoleSupported, } from "../../extensions/permissions/write/grantRole.js";
export { isRenounceRoleSupported, renounceRole, } from "../../extensions/permissions/write/renounceRole.js";
export { isRevokeRoleSupported, revokeRole, } from "../../extensions/permissions/write/revokeRole.js";
// --------------------------------------------------------
// PermissionsEnumerable
// --------------------------------------------------------
export { getAllRoleMembers, isGetAllRoleMembersSupported, } from "../../extensions/permissions/read/getAllMembers.js";
// READ
export { getRoleMember, isGetRoleMemberSupported, } from "../../extensions/permissions/read/getRoleMember.js";
export { getRoleMemberCount, isGetRoleMemberCountSupported, } from "../../extensions/permissions/read/getRoleMemberCount.js";
// --------------------------------------------------------
// Utils
// --------------------------------------------------------
export { getRoleHash, roleMap } from "../../extensions/permissions/utils.js";
