// thirdweb/dist/cjs/react/web/ui/Bridge/payment-selection/FiatProviderSelection.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FiatProviderSelection = FiatProviderSelection;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const address_js_1 = require("../../../../../utils/address.js");
const formatNumber_js_1 = require("../../../../../utils/formatNumber.js");
const units_js_1 = require("../../../../../utils/units.js");
const CustomThemeProvider_js_1 = require("../../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../../core/design-system/index.js");
const useBuyWithFiatQuotesForProviders_js_1 = require("../../../../core/hooks/pay/useBuyWithFiatQuotesForProviders.js");
const formatTokenBalance_js_1 = require("../../ConnectWallet/screens/formatTokenBalance.js");
const basic_js_1 = require("../../components/basic.js");
const buttons_js_1 = require("../../components/buttons.js");
const Img_js_1 = require("../../components/Img.js");
const Spacer_js_1 = require("../../components/Spacer.js");
const Spinner_js_1 = require("../../components/Spinner.js");
const text_js_1 = require("../../components/text.js");
const PROVIDERS = [
    {
        description: "Fast and secure payments",
        iconUri: "https://i.ibb.co/LDJ3Rk2t/Frame-5.png",
        id: "coinbase",
        name: "Coinbase",
    },
    {
        description: "Trusted payment processing",
        iconUri: "https://i.ibb.co/CpgQC2Lf/images-3.png",
        id: "stripe",
        name: "Stripe",
    },
    {
        description: "Global payment solution",
        iconUri: "https://i.ibb.co/Xx2r882p/Transak-official-symbol-1.png",
        id: "transak",
        name: "Transak",
    },
];
function FiatProviderSelection({ onProviderSelected, client, toChainId, toTokenAddress, toAddress, toAmount, currency, }) {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    // Fetch quotes for all providers
    const quoteQueries = (0, useBuyWithFiatQuotesForProviders_js_1.useBuyWithFiatQuotesForProviders)({
        amount: toAmount || "0",
        chainId: toChainId,
        client,
        currency: currency || "USD",
        receiver: (0, address_js_1.checksumAddress)(toAddress),
        tokenAddress: (0, address_js_1.checksumAddress)(toTokenAddress),
    });
    const quotes = (0, react_1.useMemo)(() => {
        return quoteQueries.map((q) => q.data).filter((q) => !!q);
    }, [quoteQueries]);
    if (quoteQueries.every((q) => q.isError)) {
        return ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { center: "both", flex: "column", style: { minHeight: "120px" }, children: (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "secondaryText", size: "sm", children: "No quotes available" }) }));
    }
    // TODO: add a "remember my choice" checkbox
    return ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "column", gap: "sm", children: quotes.length > 0 ? (quotes
            .sort((a, b) => a.currencyAmount - b.currencyAmount)
            .map((quote, index) => {
            const provider = PROVIDERS.find((p) => p.id === quote.intent.onramp);
            if (!provider) {
                return null;
            }
            return ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { animate: "fadein", style: {
                    animationDelay: `${index * 100}ms`,
                }, children: (0, jsx_runtime_1.jsx)(buttons_js_1.Button, { fullWidth: true, onClick: () => onProviderSelected(provider.id), style: {
                        backgroundColor: theme.colors.tertiaryBg,
                        border: `1px solid ${theme.colors.borderColor}`,
                        borderRadius: index_js_1.radius.md,
                        padding: `${index_js_1.spacing.sm} ${index_js_1.spacing.md}`,
                        textAlign: "left",
                    }, variant: "secondary", children: (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "row", gap: "sm", style: { alignItems: "center", width: "100%" }, children: [(0, jsx_runtime_1.jsx)(basic_js_1.Container, { style: {
                                    alignItems: "center",
                                    borderRadius: "50%",
                                    display: "flex",
                                    height: `${index_js_1.iconSize.md}px`,
                                    justifyContent: "center",
                                    overflow: "hidden",
                                    padding: index_js_1.spacing.xs,
                                    width: `${index_js_1.iconSize.md}px`,
                                }, children: (0, jsx_runtime_1.jsx)(Img_js_1.Img, { alt: provider.name, client: client, height: index_js_1.iconSize.md, src: provider.iconUri, width: index_js_1.iconSize.md }) }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "column", gap: "3xs", style: { flex: 1 }, children: (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", size: "md", style: { fontWeight: 600 }, children: provider.name }) }), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "3xs", style: { alignItems: "flex-end" }, children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", size: "sm", style: { fontWeight: 500 }, children: (0, formatTokenBalance_js_1.formatCurrencyAmount)(currency || "US", quote.currencyAmount) }), (0, jsx_runtime_1.jsxs)(text_js_1.Text, { color: "secondaryText", size: "xs", children: [(0, formatNumber_js_1.formatNumber)(Number((0, units_js_1.toTokens)(quote.destinationAmount, quote.destinationToken.decimals)), 4), " ", quote.destinationToken.symbol] })] })] }) }) }, provider.id));
        })) : ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { center: "both", flex: "column", style: { minHeight: "120px" }, children: [(0, jsx_runtime_1.jsx)(Spinner_js_1.Spinner, { color: "secondaryText", size: "lg" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, color: "secondaryText", size: "sm", children: "Generating quotes..." })] })) }));
}
