// @thirdweb-dev/react/dist/OTPInput-363a430d.cjs.dev.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var react = require('@emotion/react');
var inputOtp = require('input-otp');
var basic = require('./basic-3fc4a0fe.cjs.dev.js');
var jsxRuntime = require('react/jsx-runtime');
require('@radix-ui/colors');
require('react');
require('@emotion/styled');
require('@radix-ui/react-icons');
require('@thirdweb-dev/react-core');

function OTPInput(props) {
  return /*#__PURE__*/jsxRuntime.jsx(OTPInputContainer, {
    children: /*#__PURE__*/jsxRuntime.jsx(inputOtp.OTPInput, {
      onComplete: () => {
        props.onEnter();
      },
      maxLength: 6,
      value: props.value,
      render: _ref => {
        let {
          slots
        } = _ref;
        return /*#__PURE__*/jsxRuntime.jsx(basic.Container, {
          flex: "row",
          gap: "xs",
          center: "both",
          children: slots.map((slot, idx) => /*#__PURE__*/jsxRuntime.jsx(Slot, {
            ...slot,
            isInvalid: props.isInvalid
          }, idx))
        });
      },
      onChange: newValue => {
        props.setValue(newValue);
      }
    })
  });
}
const OTPInputContainer = /* @__PURE__ */basic.StyledDiv({
  "& input": {
    maxWidth: "100%"
  }
});
function Slot(props) {
  return /*#__PURE__*/jsxRuntime.jsxs(OTPInputBox, {
    "data-active": props.isActive,
    "data-error": props.isInvalid,
    children: [props.char !== null && /*#__PURE__*/jsxRuntime.jsx("div", {
      children: props.char
    }), props.hasFakeCaret && /*#__PURE__*/jsxRuntime.jsx(FakeCaret, {})]
  });
}
const caretBlink = react.keyframes`
  0%, 100% {
    opacity: 0;
  },
  50% {
    opacity: 1;
  }
  `;
const FakeCaret = /* @__PURE__ */basic.StyledDiv(() => {
  const theme = basic.useCustomTheme();
  return {
    position: "absolute",
    pointerEvents: "none",
    inset: 0,
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    animation: `${caretBlink} 1s infinite`,
    "&::before": {
      content: "''",
      display: "block",
      width: "1.5px",
      height: "1em",
      backgroundColor: theme.colors.primaryText
    }
  };
});
const OTPInputBox = /* @__PURE__ */basic.StyledDiv(() => {
  const theme = basic.useCustomTheme();
  return {
    position: "relative",
    width: "40px",
    height: "40px",
    textAlign: "center",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    fontSize: basic.fontSize.md,
    padding: basic.spacing.xs,
    boxSizing: "border-box",
    transition: "color 200ms ease",
    border: `2px solid ${theme.colors.borderColor}`,
    "&[data-active='true']": {
      borderColor: theme.colors.accentText
    },
    color: theme.colors.primaryText,
    borderRadius: basic.radius.lg,
    "&[data-error='true']": {
      borderColor: theme.colors.danger
    }
  };
});

exports.OTPInput = OTPInput;
