// thirdweb/dist/esm/stories/Bridge/TransactionPayment.stories.js
import { jsx as _jsx } from "react/jsx-runtime";
import { TransactionPayment, } from "../../react/web/ui/Bridge/TransactionPayment.js";
import { ModalThemeWrapper, storyClient } from "../utils.js";
import { TRANSACTION_UI_OPTIONS } from "./fixtures.js";
// Wrapper component to provide theme context
const TransactionPaymentWithTheme = (props) => {
    const { theme, ...componentProps } = props;
    return (_jsx(ModalThemeWrapper, { theme: theme, children: _jsx("div", { style: { width: "400px" }, children: _jsx(TransactionPayment, { ...componentProps }) }) }));
};
const meta = {
    args: {
        client: storyClient,
        onExecuteTransaction: () => { },
        onContinue: (_amount, _token, _receiverAddress) => { },
        theme: "dark",
        uiOptions: TRANSACTION_UI_OPTIONS.ethTransfer,
    },
    argTypes: {
        onContinue: {
            action: "continue clicked",
            description: "Called when user continues with the transaction",
        },
        theme: {
            control: "select",
            description: "Theme for the component",
            options: ["light", "dark"],
        },
        uiOptions: {
            description: "UI configuration for transaction mode including prepared transaction",
        },
    },
    component: TransactionPaymentWithTheme,
    parameters: {
        docs: {
            description: {
                component: "Transaction payment component that displays detailed transaction information including contract details, function names, transaction costs, and network fees.\n\n" +
                    "## Features\n" +
                    "- **Contract Information**: Shows contract name and clickable address\n" +
                    "- **Function Detection**: Extracts function names from transaction data using ABI\n" +
                    "- **Cost Calculation**: Displays transaction value and USD equivalent\n" +
                    "- **Network Fees**: Shows estimated gas costs with token amounts\n" +
                    "- **Chain Details**: Network name and logo with proper formatting\n" +
                    "- **Skeleton Loading**: Comprehensive loading states matching final layout\n\n" +
                    "This component now accepts uiOptions directly to configure the transaction and metadata. Supports both native token and ERC20 token transactions with proper function name extraction.",
            },
        },
        layout: "centered",
    },
    tags: ["autodocs"],
    title: "Bridge/TransactionPayment",
};
export default meta;
export const EthereumTransfer = {
    args: {
        theme: "dark",
        uiOptions: TRANSACTION_UI_OPTIONS.ethTransfer,
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Simple ETH transfer transaction showing native token value and network fees with USD conversion. Demonstrates function name extraction from contract ABI.",
            },
        },
    },
};
export const EthereumTransferLight = {
    args: {
        theme: "light",
        uiOptions: TRANSACTION_UI_OPTIONS.ethTransfer,
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "Simple ETH transfer transaction in light theme with skeleton loading support.",
            },
        },
    },
};
export const ERC20TokenTransfer = {
    args: {
        theme: "dark",
        uiOptions: TRANSACTION_UI_OPTIONS.erc20Transfer,
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "ERC20 token transaction showing token amount, USD value, and proper formatting using real token data. Displays transfer function details.",
            },
        },
    },
};
export const ERC20TokenTransferLight = {
    args: {
        theme: "light",
        uiOptions: TRANSACTION_UI_OPTIONS.erc20Transfer,
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "ERC20 token transaction in light theme with enhanced formatting.",
            },
        },
    },
};
export const ContractInteraction = {
    args: {
        theme: "dark",
        uiOptions: TRANSACTION_UI_OPTIONS.contractInteraction,
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Complex contract interaction showing function name extraction from ABI (claimTo), cost calculation, and network details with proper currency formatting.",
            },
        },
    },
};
export const ContractInteractionLight = {
    args: {
        theme: "light",
        uiOptions: TRANSACTION_UI_OPTIONS.contractInteraction,
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "Contract interaction transaction in light theme with enhanced UX and skeleton loading.",
            },
        },
    },
};
export const CustomButtonLabel = {
    args: {
        theme: "dark",
        uiOptions: TRANSACTION_UI_OPTIONS.customButton,
    },
    parameters: {
        backgrounds: { default: "dark" },
        docs: {
            description: {
                story: "Example showcasing custom button label functionality. The button shows 'Execute Now' instead of the default 'Execute [functionName]' text.",
            },
        },
    },
};
export const CustomButtonLabelLight = {
    args: {
        theme: "light",
        uiOptions: TRANSACTION_UI_OPTIONS.customButton,
    },
    parameters: {
        backgrounds: { default: "light" },
        docs: {
            description: {
                story: "Light theme version with custom button label 'Execute Now'.",
            },
        },
    },
};
